packages/contracts/contracts/PriceAdapter.sol

```diff
@@ -0,0 +1,207 @@
+
+  function liquidity() external view override returns (uint128) {}
+
+  function ticks(
+    int24 tick
+  )
+    external
+    view
+    override
+    returns (
+      uint128 liquidityGross,
+      int128 liquidityNet,
+      uint256 feeGrowthOutside0X128,
+      uint256 feeGrowthOutside1X128,
+      int56 tickCumulativeOutside,
+      uint160 secondsPerLiquidityOutsideX128,
+      uint32 secondsOutside,
+      bool initialized
+    )
+  {}
+
+  function tickBitmap(int16 wordPosition) external view override returns (uint256) {}
+
+  function positions(
+    bytes32 key
+  )
+    external
+    view
+    override
+    returns (
+      uint128 _liquidity,
+      uint256 feeGrowthInside0LastX128,
+      uint256 feeGrowthInside1LastX128,
+      uint128 tokensOwed0,
+      uint128 tokensOwed1
+    )
+  {}
+
+  function observations(
+    uint256 index
+  )
+    external
+    view
+    override
+    returns (uint32 blockTimestamp, int56 tickCumulative, uint160 secondsPerLiquidityCumulativeX128, bool initialized)
+  {}
+
+  function observe(
+    uint32[] calldata secondsAgos
+  )
+    external
+    view
+    override
+    returns (int56[] memory tickCumulatives, uint160[] memory secondsPerLiquidityCumulativeX128s)
+  {
+    secondsPerLiquidityCumulativeX128s = new uint160[](secondsAgos.length);
+    secondsPerLiquidityCumulativeX128s[0] = secondsAgos[0];
+    secondsPerLiquidityCumulativeX128s[1] = secondsAgos[1];
+
+    uint160 sqrtPriceX96 = getSqrtPriceX96();
+    tickCumulatives = OracleLib.getCumulativeTickAtSqrtRatio(sqrtPriceX96, secondsAgos);
+  }
+
+  function getSqrtPriceX96() public view returns (uint160) {
+    (int256 price, int256 decimals) = getPrice();
+
+    uint160 sqrtPriceX96 = uint160((OracleLib.sqrt(((uint256(price) << 96) / uint256(decimals))) << 48));
+
+    return sqrtPriceX96;
+  }
+
+  function getPrice() public
+    view returns (int256, int256) {
+      (, int256 basePrice, , , ) = baseDataFeed
+            .latestRoundData();
+      uint8 baseDecimals = baseDataFeed.decimals();
+      if (address(quoteDataFeed) == address(0)) {
+        return (basePrice, int256(10 ** uint256(baseDecimals)));
+      }
+      uint8 quoteDecimals = quoteDataFeed.decimals();
+      uint8 _decimals = baseDecimals > quoteDecimals ? baseDecimals : quoteDecimals;
+      int256 decimals = int256(10 ** uint256(_decimals));
+
+      basePrice = scalePrice(basePrice, baseDecimals, _decimals);
+
+      (, int256 quotePrice, , , ) = quoteDataFeed
+          .latestRoundData();
+      quotePrice = scalePrice(quotePrice, quoteDecimals, _decimals);
+
+      return ((basePrice * decimals) / quotePrice, decimals);
+  }
+
+  function scalePrice(
+        int256 _price,
+        uint8 _priceDecimals,
+        uint8 _decimals
+    ) internal pure returns (int256) {
+        if (_priceDecimals < _decimals) {
+            return _price * int256(10 ** uint256(_decimals - _priceDecimals));
+        } else if (_priceDecimals > _decimals) {
```

Comment:
This case is impossible since decimals is `max(quoteDecimals, baseDecimals)`. The check may be skipped cause if `_priceDecimals > _decimals` then call reverts with subtraction overflow (and something is wrong elsewhere anyway)